File: src/GameRenderer.cpp
#include "GameRenderer.hpp"

#include "components/Camera.hpp"
#include "engine/entity-system/include.hpp"
#include "engine/game-loop/SingleThreadGameLoop.hpp"
#include "engine/resource-system/include.hpp"
#include "engine/utils/timing/print-fps.hpp"
#include "render.hpp"
#include "Settings.hpp"

#include "engine/utils/debug/xtrace.hpp"

GameRenderer::GameRenderer(ComponentManager& manager, ResourceStorage& storage)
 : componentManager(manager),
   resourceStorage(storage) {
    Camera& camera = storage.get<Camera>("camera");
    Settings& settings = storage.get<Settings>("settings");

    windowPtr = std::make_unique<sf::RenderWindow>(
        sf::VideoMode(camera.width, camera.height),
        settings.getWindowTitle()
    );

    adjustView();
}

void GameRenderer::operator()(SingleThreadGameLoop& game) {
    engine::utils::printFPS<2>("FPS", 2000);
    // static Settings& settings = resourceStorage.get<Settings>("settings");
    // static int minWindowWidth = settings.getMinWindowWidth();
    // static int minWindowHeight = settings.getMinWindowHeight();

    adjustView();

    sf::RenderWindow& window = *windowPtr;

    sf::Event event;
    while (window.pollEvent(event)) {
        if (event.type == sf::Event::Closed) {
            game.stop();
        } else if (event.type == sf::Event::Resized) {
            sf::Vector2u windowSize = window.getSize();
            // XTRACE(windowSize.x);
            // XTRACE(windowSize.y);
            // XTRACE(minWindowWidth);
            // XTRACE(minWindowHeight);
            // bool adjustedSize = false;

            // if (windowSize.x < minWindowWidth) {
            //     windowSize.x = minWindowWidth;
            //     adjustedSize = true;
            // }

            // if (windowSize.y < minWindowHeight) {
            //     windowSize.y = minWindowHeight;
            //     adjustedSize = true;
            // }

            // if (adjustedSize) {
            //     window.setSize(windowSize);
            //     ECHO("[CHANGE]");
            //     XTRACE(windowSize.x);
            //     XTRACE(windowSize.y);
            //     XTRACE(minWindowWidth);
            //     XTRACE(minWindowHeight);
            //     ECHO("------------");
            // }

            Camera& camera = resourceStorage.get<Camera>("camera");
            camera.width = windowSize.x;
            camera.height = windowSize.y;
            adjustView();
        }
    }

    window.clear();
    render(window, componentManager, resourceStorage);
    window.display();
}

void GameRenderer::adjustView() {
    Camera& camera = resourceStorage.get<Camera>("camera");
    view.reset(sf::FloatRect(camera.x, camera.y, camera.width, camera.height));
    windowPtr->setView(view);
}

File: frontend/src/GameRenderer.ts
import { Camera } from './Camera';
import { GameLoop } from './engine/GameLoop';
import { ResourceStorage } from './engine/resource-system/ResourceStorage';
import { render } from './render';

export class GameRenderer {
    constructor(
        private resourceStorage: ResourceStorage,
        private canvas: HTMLCanvasElement
    ) {
        const camera = resourceStorage.get<Camera>('camera');
        canvas.width = camera.width;
        canvas.height = camera.height;
        this.adjustView();

        this.camera = camera;
        this.context = this.getContext();
    }

    render(gameLoop: GameLoop): void {
        this.adjustView();
        this.clearCanvas();
        render(this.context, this.resourceStorage);
    }

    private getContext(): CanvasRenderingContext2D {
        const context = this.canvas.getContext('2d');

        if (!context) {
            throw Error('Failed to get 2D context');
        }

        return context;
    }

    private adjustView(): void {
        // TODO
    }

    private clearCanvas(): void {
        this.context.clearRect(0, 0, this.camera.width, this.camera.height);
    }

    private camera: Camera;
    private context: CanvasRenderingContext2D;
}
